core/src/main/java/org/apache/druid/timeline/partition/BaseDimensionRangeShardSpec.java

```diff
@@ -0,0 +1,109 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.timeline.partition;
+
+import com.google.common.collect.Ordering;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.StringTuple;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.guava.Comparators;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+
+public abstract class BaseDimensionRangeShardSpec implements ShardSpec
+{
+  protected final List<String> dimensions;
+  @Nullable
+  protected final StringTuple start;
+  @Nullable
+  protected final StringTuple end;
+
+  protected BaseDimensionRangeShardSpec(
+      List<String> dimensions,
+      @Nullable StringTuple start,
+      @Nullable StringTuple end
+  )
+  {
+    this.dimensions = dimensions;
+    this.start = start;
+    this.end = end;
+  }
+
+  @Override
+  public ShardSpecLookup getLookup(final List<? extends ShardSpec> shardSpecs)
+  {
+    return createLookup(dimensions, shardSpecs);
+  }
+
+  private static ShardSpecLookup createLookup(List<String> dimensions, List<? extends ShardSpec> shardSpecs)
+  {
+    BaseDimensionRangeShardSpec[] rangeShardSpecs = new BaseDimensionRangeShardSpec[shardSpecs.size()];
+    for (int i = 0; i < shardSpecs.size(); i++) {
+      rangeShardSpecs[i] = (BaseDimensionRangeShardSpec) shardSpecs.get(i);
+    }
+    final Comparator<StringTuple> startComparator = Comparators.naturalNullsFirst();
+    final Comparator<StringTuple> endComparator = Ordering.natural().nullsLast();
+    final Comparator<BaseDimensionRangeShardSpec> shardSpecComparator = new Comparator<BaseDimensionRangeShardSpec>()
+    {
+      @Override
+      public int compare(BaseDimensionRangeShardSpec o1, BaseDimensionRangeShardSpec o2)
+      {
+        int startComparison = startComparator.compare(o1.start, o2.start);
+        if (startComparison != 0) {
+          return startComparison;
+        }
+        return endComparator.compare(o1.end, o2.end);
+      }
+    };
+    Arrays.sort(rangeShardSpecs, shardSpecComparator);
+
+    return (long timestamp, InputRow row) -> {
+      StringTuple inputRowTuple = getInputRowTuple(dimensions, row);
+      int startIndex = 0;
+      int endIndex = shardSpecs.size() - 1;
+      while (startIndex <= endIndex) {
+        int mid = (startIndex + endIndex) >>> 1;
+        BaseDimensionRangeShardSpec rangeShardSpec = rangeShardSpecs[mid];
+        if (startComparator.compare(inputRowTuple, rangeShardSpec.start) < 0) {
+          endIndex = mid - 1;
+        } else if (endComparator.compare(inputRowTuple, rangeShardSpec.end) < 0) {
+          return rangeShardSpec;
+        } else {
+          startIndex = mid + 1;
+        }
+      }
```

Comment:
The array component is BaseDimensionRangeShardSpec while the key is StringTuple, so I can not directly call Arrays.binarySearch

Reply:
You could use a dummy key. It would still be cleaner than writing the binary search logic yourself.
Something like this:

```java
final StringTuple searchTuple = getInputRowTuple(dimensions, row);
final BaseDimensionRangeShardSpec searchKey = new DimensionRangeShardSpec(dimensions, searchTuple, searchTuple, 0, 1);
final int searchResult = Arrays.binarySearch(rangeShardSpecs, searchKey, shardSpecComparator);
if (searchResult < 0) {
   throw new ISE("row[%s] doesn't fit in any shard[%s]", row, shardSpecs);
} else {
    return rangeShardSpecs[searchResult];
}
```

Please let me know if this seems cleaner.

Reply:
Arrays.binarySearch requires the searchKey to equals one array element, while actually we do need find one that equals or contains the searchKey. For example:

input:
```
searchKey: ["a","a"]
shardSpecs:[[null,"c"], ["c", "h"],["h",null]]
```

then  the expect result is [null, "c"], but the Arrays.binarySearch will return -1, not 0,  which means ["a","a"] <= [null, "c"].